[mlir][spirv] Add mgpu* wrappers for Vulkan runtime, migrate some tests

andfau-amd opened the pull request:
This commit adds new wrappers around the MLIR Vulkan runtime which implement the mgpu* APIs (as generated by GPUToLLVMConversionPass), adds an optional LLVM lowering to the Vulkan runner mlir-opt pipeline based on GPUToLLVMConversionPass, and migrates several of the mlir-vulkan-runner tests to use mlir-cpu-runner instead, together with the new pipeline and wrappers.

This is a further incremental step towards eliminating mlir-vulkan-runner and its associated pipeline, passes and wrappers (#73457). This commit does not migrate all of the tests to the new system, because changes to the mgpuLaunchKernel ABI will be necessary to support the tests that use multi-dimensional memref arguments.

Hardcode84 commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -91,6 +123,88 @@ void bindMemRef(void *vkRuntimeManager, DescriptorSetIndex setIndex,
+}
+
+VULKAN_WRAPPER_SYMBOL_EXPORT void mgpuStreamDestroy(void *vkRuntimeManager) {
+  delete reinterpret_cast<VulkanRuntimeManager *>(vkRuntimeManager);
```
nit: you can use `static_cast` when casting pointers to/from `void*`

Hardcode84 commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -26,6 +28,36 @@
+  ~VulkanModule() = default;
+
+  VulkanFunction *getFunction(const char *name) {
+    return &functions.emplace_back(this, name);
```
vector reallocation on `emplace_back` will invalidate previously obtained pointers.

andfau-amd replied:
Ah, good point! Does `std::vector<std::unique_ptr<VulkanFunction>>` seem like a better idea then?

Hardcode84 replied:
Yeah, this should work

andfau-amd commented on mlir/include/mlir/Conversion/Passes.td:
```diff
@@ -517,6 +517,11 @@ def GpuToLLVMConversionPass : Pass<"gpu-to-llvm", "ModuleOp"> {
+    Option<"typeCheckKernelArgs", "type-check-kernel-args", "bool",
+           /*default=*/"false",
+             "Require all kernel arguments to be memrefs of rank 1 and with a "
+             "32-bit element size."
```
It is temporary indeed, but I didn't want to repeat the explanation from circa line 979 in the implementation. I am hoping to remove this already in the next commit.

andfau-amd commented on mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:
```diff
@@ -966,6 +969,26 @@ LogicalResult LegalizeLaunchFuncOpPattern::matchAndRewrite(
+    // which takes an untyped list of arguments. The type check here prevents
+    // accidentally violating the assumption made in vulkan-runtime-wrappers.cpp
+    // and creating a unchecked runtime ABI mismatch.
+    // TODO: Change the ABI here to remove the need for this type check.
```
Can I link to https://github.com/llvm/llvm-project/issues/73457 or do I need a more specific issue?

andfau-amd commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -26,6 +28,37 @@
+  uint32_t blobSize() { return static_cast<uint32_t>(blob.size()); }
+
+private:
+  std::vector<char> blob;
+  std::vector<std::unique_ptr<VulkanFunction>> functions;
```
Actually I seem to be wrong here, there's already some use of the LLVM Support libraries. I guess I can try using a SmallVector.

andfau-amd replied:
Actually no, I think I was right the first time:

https://github.com/llvm/llvm-project/blob/9e863cd44945345f22a28cdd3ea12aaa7963345e/mlir/tools/mlir-vulkan-runner/CMakeLists.txt#L42-L49

The usage of the Support libraries must be in some header-only fashion. I guess everything about `SmallVector` is probably templated but I'd still rather avoid using it, it feels risky.

kuhar commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -26,6 +28,37 @@
+  }
+
+  uint8_t *blobData() { return reinterpret_cast<uint8_t *>(blob.data()); }
+  uint32_t blobSize() { return static_cast<uint32_t>(blob.size()); }
```
Not necessarily because it's common to pass blobs around as `char*` or equivalent to avoid strict aliasing violations. And if we do stick with bytes, this would be surprising on its own given how the Vulkan api uses dwords: https://registry.khronos.org/vulkan/specs/latest/man/html/VkShaderModuleCreateInfo.html

kuhar replied:
Ah sorry, it's even weirder: Vulkan expects you to pass data as dword* but the code size is the number of bytes:
> codeSize is the size, in bytes, of the code pointed to by pCode.

So in any case, please add a unit to make this clear.

andfau-amd replied:
I don't want to add units because I think it creates a false sense of security. You can only really tell if the code is correct by auditing the whole chain, and what the code actually does and the types it uses are a stronger signal than names or comments. The implementation of `VulkanModule` speaks for itself.

andfau-amd replied:
To make things clearer though, I've decided to remove one of the type casts (not sure why I had it) and move the other.

kuhar replied:
> what the code actually does and the types it uses are a stronger signal than names or comments

I agree with this but this is not an all or nothing approach in absence of a dedicated unit type. Having worked with similar code in the past across multiple projects, I've learned to be extra cautious around how we represent SPIR-V modules. Also note that it's quite common to see guidelines around time units in many projects, e.g.:
* https://users.ece.cmu.edu/~eno/coding/CppCodingStandard.html#units
* https://ruudvanasseldonk.com/2022/03/20/please-put-units-in-names
* https://github.com/search?q=repo%3Airee-org%2Firee+inbytes+&type=code.

At google, we had an internal lint that would complain about any variables of primitive types named something like `duration` or `time` without an appropriate suffix.

andfau-amd replied:
I guess it can't hurt. Is it more to your liking now?

kuhar commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -91,6 +124,91 @@ void bindMemRef(void *vkRuntimeManager, DescriptorSetIndex setIndex,
+  //       that other memref types can also be used. This will allow migrating
+  //       the remaining tests and removal of mlir-vulkan-runner.
+  const size_t paramsPerMemRef = 5;
+  if (paramsCount % paramsPerMemRef) {
+    abort();
```
```suggestion
  if (paramsCount % paramsPerMemRef != 0) {
    abort();
```
Should we also print some error message?

andfau-amd replied:
This is more or less an assert for something that "shouldn't happen", and I'm hoping the automatic backtrace is sufficient.

kuhar commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -91,6 +124,91 @@ void bindMemRef(void *vkRuntimeManager, DescriptorSetIndex setIndex,
+  for (size_t i = 0; i < paramsCount; i += paramsPerMemRef) {
+    auto memref = static_cast<MemRefDescriptor<uint32_t, 1> *>(params[i]);
+    bindMemRef<uint32_t, 1>(manager, setIndex, bindIndex, memref);
+    bindIndex++;
```
nit: https://llvm.org/docs/CodingStandards.html#prefer-preincrement

```suggestion
    ++bindIndex;
```

kuhar reviewed (approved):
LGTM, thanks for the fixes

Hardcode84 commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -26,6 +28,38 @@
+  }
+
+  uint8_t *blobData() { return blob.data(); }
+  size_t blobSizeInBytes() { return blob.size(); }
```
nit: const

andfau-amd replied:
I don't think I can make these use `const` (unless you mean only the qualifier on the method?) without changing the `VulkanRuntime` code to accept `const` pointers. I don't think there's any particularly good reason for it not taking a `const`-qualified pointer, but I don't want to touch that code right now.

Hardcode84 replied:
yeah, I mean `size_t blobSizeInBytes() const { return blob.size(); }`

andfau-amd committed:
[mlir][spirv] Add mgpu* wrappers for Vulkan runtime, migrate some tests

This commit adds new wrappers around the MLIR Vulkan runtime which
implement the mgpu* APIs (as generated by GPUToLLVMConversionPass),
adds an optional LLVM lowering to the Vulkan runner mlir-opt pipeline
based on GPUToLLVMConversionPass, and migrates several of the
mlir-vulkan-runner tests to use mlir-cpu-runner instead, together with
the new pipeline and wrappers.

This is a further incremental step towards eliminating
mlir-vulkan-runner and its associated pipeline, passes and wrappers
(#73457). This commit does not migrate all of the tests to the new
system, because changes to the mgpuLaunchKernel ABI will be necessary
to support the tests that use multi-dimensional memref arguments.
